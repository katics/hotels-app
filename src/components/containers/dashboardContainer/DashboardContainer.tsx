import { FC } from "react";

import store from "../../../store";
import { hotelsSelector } from "../../../store/hotel/hotelSelector";
import HotelProvider, {
  useDispatchHotel,
  useSelectorHotel,
  hotelContext,
} from "../../../store/HotelProvider";

import {
  addRemoveUserFromFav,
  fetchHotels,
  favouriteHotels,
} from "../../../store/hotel/hotelActions";
import { userSeletor } from "../../../store/user/userSelector";
import DashboardLayout from "../../layout/dashboardLayout/DashboardLayout";
import Spinner from "../../spinner/Spinner";
import { FavouriteHotelRequest } from "../../../utils/types/FavouriteHotelRequest";
import { Hotel } from "../../../utils/types/Hotel";

const DashboardPage: FC = () => (
  <HotelProvider store={store} context={hotelContext}>
    <DashboardContainer />
  </HotelProvider>
);

const DashboardContainer: FC = () => {
  const fetchHotelForCurrentUser = () => {
    hotelDispatch(fetchHotels());
  };

  const { listOfHotels, isLoading } = useSelectorHotel(hotelsSelector);
  const { currentUser } = useSelectorHotel(userSeletor);
  const hotelDispatch = useDispatchHotel();

  const addRemoveUserFromHotels = (hotelid: number, userId: number) => {
    const hotels = listOfHotels.map(hotel => {
      if (hotel.id === hotelid) {
        if (hotel.user.includes(userId)) {
          hotel.user = hotel.user.filter(user => {
            user !== userId;
          });
        } else {
          hotel.user.push(userId);
        }
      }
      return hotel;
    });
    hotelDispatch(addRemoveUserFromFav(hotels));
  };

  const addRemoveFavHotel = (hotelId: number, isFavorite: boolean) => {
    const favHotel: FavouriteHotelRequest = {
      hotel_id: hotelId,
      is_favorite: isFavorite,
    };
    addRemoveUserFromHotels(hotelId, currentUser.user_id);
    hotelDispatch(favouriteHotels(favHotel));
  };
  return (
    <>
      {!isLoading ? (
        <DashboardLayout
          hotels={listOfHotels}
          title="List of all hotels"
          fetchHotels={() => fetchHotelForCurrentUser()}
          user={currentUser}
          addRemoveFavHotel={addRemoveFavHotel}
        />
      ) : (
        <Spinner />
      )}
    </>
  );
};

export default DashboardPage;
