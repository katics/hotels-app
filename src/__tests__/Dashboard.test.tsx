import DashboardContainer from "../components/containers/dashboardContainer/DashboardContainer";
import { StaticRouter } from "react-router-dom";

import { render } from "@testing-library/react";

describe("Dashboard Page", () => {
  it("Should match dashboard page snapshot", () => {
    const { container } = render(
      <StaticRouter>
        <DashboardContainer />
      </StaticRouter>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
