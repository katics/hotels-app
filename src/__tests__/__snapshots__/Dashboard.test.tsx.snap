// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard Page Should match dashboard page snapshot 1`] = `
<div
  class="col-md-12 col-lg-12"
>
  <div
    class="row"
  >
    <div>
      <button
        class="btn btn-primary"
      >
        List of all hotels
      </button>
    </div>
  </div>
</div>
`;
